torch/csrc/jit/codegen/cuda/lower_allocation.cpp

```diff
@@ -0,0 +1,357 @@
+      size = ir_builder.create<kir::Int>(1);
+    } else {
+      size = alloc_dims[0];
+      for (size_t i = 1; i < alloc_dims.size(); i++) {
+        size = ir_builder.mulExpr(size, alloc_dims[i]);
+      }
+    }
+
+    // Create the allocation node
+    info.alloc_expr = ir_builder.create<kir::Allocate>(
+        info.buffer, info.buffer->memoryType(), size);
+  }
+
+  void handle(kir::Expr* expr) {
+    if (!ir_utils::isTVOp(expr) || expr->isA<kir::Allocate>()) {
+      expr->accept(this);
+      return;
+    }
+
+    // // Found where the allocation needs to be inserted
+
+    for (auto out : expr->outputs()) {
+      if (!out->isA<kir::TensorView>()) {
+        continue;
+      }
+
+      auto out_tv = out->as<kir::TensorView>();
+
+      kir::Val* init = nullptr;
+      if (expr->isA<kir::ReductionOp>() && out_tv->fuserTv()->hasReduction()) {
+        init = expr->as<kir::ReductionOp>()->init();
+      }
+
+      const bool is_output = std::find(
+                                 gpu_lower->kernel()->outputs().begin(),
+                                 gpu_lower->kernel()->outputs().end(),
+                                 out) != gpu_lower->kernel()->outputs().end();
+
+      // Don't need to alloc outputs, and if we don't need to initialize we're
+      // done.
+      if (is_output && init == nullptr) {
+        continue;
+      }
+
+      AllocationInformation allocation;
+      allocation.buffer = out_tv;
+      findAllocationPosition(allocation, expr);
+      createAllocExpr(allocation, is_output);
+      createInitExpr(allocation, init);
+
+      allocs.push_back(allocation);
+    }
+  }
+
+  void visit(kir::ForLoop* fl) final {
+    for_loops.push_back(fl);
+    // Modifying in place, make a copy of the vector
+    const std::vector<kir::Expr*> exprs = fl->body().exprs();
+    for (auto expr : exprs) {
+      handle(expr);
+    }
+    for_loops.pop_back();
+  }
+
+  void visit(kir::IfThenElse*) final {
+    TORCH_INTERNAL_ASSERT(
+        false,
+        "Pass does not support conditional statements, ",
+        "this pass should be run before any conditionals are placed in code.");
+  }
+
+  AllocationInserter(std::vector<kir::Expr*> _loop_nests)
+      : loop_nests_(std::move(_loop_nests)),
+        gpu_lower(GpuLower::current()),
+        ir_builder(gpu_lower->kernel()) {
+    // Compute all allocations
+    const std::vector<kir::Expr*> exprs = loop_nests_;
+    for (auto expr : exprs) {
+      handle(expr);
+    }
+
+    // Place allocations in topological order
+    auto dynamic_smem_insertion_point = loop_nests_.begin();
+    for (const auto& alloc : allocs) {
+      if (alloc.alloc_expr == nullptr) {
+        continue;
+      }
+      // Dynamic smem exprs need to be at the begining of the kernel outside for
+      // loops
+      if (alloc.buffer->memoryType() == MemoryType::Shared &&
+          !kir::ExpressionEvaluator::isConst(alloc.alloc_expr->size())) {
+        dynamic_smem_insertion_point =
```

Comment:
Does this stay valid with the loop_nests_.insert on line 301?

Reply:
Ah, that's probably not. Will update. Thanks for the feedback.

Reply:
This is fixed.